### ISSUE TITLE
Job events on different machines

### ISSUE THREAD

**Issue by gitadept** [agenda/agenda#486]
Hi! Couldn't find similar issue. Sorry if it is already asked.
I would like to see following scenario:

Machine1 creates jobs
Machine2 processes jobs
Machine1 listens for queue events like 'complete', 'success' or 'fail'.

Is this possible?
Thanks

**Comment by emhagman** [comment]
@gitadept We currently do this at my company. This is not supported by Agenda unfortunately. The event of job completion, etc, only occurs on the machine the job completes on.

What we did is use redis pub/sub to resolve (or reject) promises when the job completes. If you are interested in more details, I can go into it more but the pseudo code is:

1. `startRemoteJob` creates a promise with UUID generated as _id, stored on the Agenda job (Agenda.now, Agenda.schedule, etc is called in here).
2. The machine the job starts on has a callback assigned in object `RESULTS_PROMISE` with key of _id. When this function is called, it resolves or rejects the promise created above (the promise has a reference to the function because of closure)
3. All running machines listen to Agenda job completion events (locally)
4. When the job completes, fails, etc, a Redis pub message is created on a channel telling everyone listening what happened (completed, failed, etc).
5. The code listening for a result (via Agenda events) checks if the `_id` has a callback on the machine. If so, it will call the callback that is stored in the `RESULTS_PROMISE`, resolving or rejecting the promise. Otherwise, it ignores the message.

**Comment by OmgImAlexis** [comment]
@emhagman could you not the processing server listening to the MongoDB oplog so it can react to changes to documents such as when a job finishes?

**Comment by emhagman** [comment]
Yup you could definitely do this as well :) @OmgImAlexis That's probably a much better idea. If I get around to it, I can test it out in our code base.
